### PR TITLE
Accept GitHub issues numbered only 32426 or above

### PR DESCRIPTION
**Pull request by menkotoglou**
Fixes https://github.com/python/core-workflow/issues/504.

Iterates through the template file to find the gh-issue line and checks number if it's in the accepted issue number range.

This is how an unsuccessful attempt looks like:

```sh
❯ /bin/python3 /path/core-workflow/blurb/blurb.py

Error: The gh-issue number should be 32426 or above.

[Hit return to retry (or Ctrl-C to abort)>
```

### CODE REVIEW

File: blurb/blurb.py [hunk 1]
```diff
@@ -956,9 +956,13 @@ def init_tmp_with_template():
             failure = str(e)
 
         if not failure:
-            assert len(blurb) # if parse_blurb succeeds, we should always have a body
-            if len(blurb) > 1:
-                failure = "Too many entries!  Don't specify '..' on a line by itself."
```
Comment (ezio-melotti) [inline blurb/blurb.py:961]: Should these checks be kept too?

File: blurb/blurb.py [hunk 2]
```diff
@@ -956,9 +956,13 @@ def init_tmp_with_template():
...
+            with open(tmp_path, "rt", encoding="utf-8") as file:
```
Comment (ezio-melotti) [inline blurb/blurb.py:959]: The file seems to be already read in `blurb.load/parse`, so you should be able to access it (or even the issue number directly) directly from the instance without opening and parsing the file again.

File: blurb/blurb.py [hunk 3]
```diff
@@ -956,9 +956,13 @@ def init_tmp_with_template():
...
+                for line in file:
+                    if line.startswith(".. gh-issue:"):
+                        issue_number = line.split(":")[1].strip()
+                        if issue_number.isdigit() and int(issue_number) < 32426:
```
Comment (ezio-melotti) [inline blurb/blurb.py:963]: Should this also handle (and reject) non-numeric issue numbers?

Comment (menkotoglou) [reply blurb/blurb.py:963]: Hey @ezio-melotti, that's true. I'll split the two if statements to show different errors.

File: blurb/blurb.py [hunk 4]
```diff
@@ -956,9 +956,14 @@ def init_tmp_with_template():
...
         if not failure:
-            assert len(blurb) # if parse_blurb succeeds, we should always have a body
+            assert len(blurb)  # if parse_blurb succeeds, we should always have a body
             if len(blurb) > 1:
                 failure = "Too many entries!  Don't specify '..' on a line by itself."
+            gh_issue_number = blurb[0][0]["gh-issue"]
+            if not gh_issue_number.isdigit():
+                failure = "The gh-issue number must be a number."
+            if int(gh_issue_number) < 32426:
+                failure = "The gh-issue number should be 32426 or above."
```
Comment (hugovk) [inline blurb/blurb.py:966]: Bit of whitespace for readability, and wording:
```suggestion
            gh_issue_number = blurb[0][0]["gh-issue"]

            if not gh_issue_number.isdigit():
                failure = "The gh-issue number must be a number."

            if int(gh_issue_number) < 32426:
                failure = "The gh-issue number must be 32426 or above."
```

**Review by hugovk — COMMENTED** [review]
Please could you add some tests?

Hopefully we can get https://github.com/python/core-workflow/pull/520 merged soon, then we can add unit tests.

In the meantime, tests are done by adding valid blurb files in the `pass` dir and invalid ones in the `fail` dir at https://github.com/python/core-workflow/tree/main/blurb/tests, and then running `blurb test`.

File: blurb/blurb.py (hunk 4 above)
Comment (menkotoglou) [reply blurb/blurb.py]: Removed completely from here since we should be raising an exception instead of displaying the failure.

**Comment by menkotoglou** [comment]
Seems like the `not_number` case is already covered in the code. That's validated in the pre-existing `invalid-gh-number.rst` test case. 

Also, the tests wouldn't pass if we didn't raise an exception at the `class Blurbs` level. Decided to move to the class definition level as I'm aligned with this approach.

Finally, retrospectively fixed some old tests that indeed were using smaller GH issue numbers.

File: blurb/blurb.py [hunk 5]
```diff
@@ -486,6 +486,9 @@ def finish_entry():
                 # we'll complain about the *first* error
                 # we see in the blurb file, which is a
                 # better user experience.
+                if key == "gh-issue" and int(value) < 32426:
+                    throw("The gh-issue number should be 32426 or above.")
```
Comment (hugovk) [inline blurb/blurb.py:490]: Maybe something like this? It gives a hint they may be using the PR number instead of issue number.

```suggestion
                    throw("The gh-issue number must be 32426 or above, not a PR number.")
```

**Review by hugovk — COMMENTED** [review]
Thanks! It's nice when adding tests reveals a problem in the original implementation :)

Looks good, here's a wording suggestion.

**Comment by menkotoglou** [comment]
Thanks a lot for the review @hugovk. Suggestion commited!

**Review by hugovk — APPROVED** [review]
Thanks!

File: blurb/blurb.py [hunk 6]
```diff
@@ -956,7 +959,7 @@ def init_tmp_with_template():
             failure = str(e)
 
         if not failure:
-            assert len(blurb) # if parse_blurb succeeds, we should always have a body
+            assert len(blurb)  # if parse_blurb succeeds, we should always have a body
```
Comment (pradyunsg) [inline blurb/blurb.py:962]: nit: unrelated whitespace change.

File: blurb/blurb.py [hunk 7]
```diff
@@ -486,6 +486,9 @@ def finish_entry():
                 # we'll complain about the *first* error
                 # we see in the blurb file, which is a
                 # better user experience.
+                if key == "gh-issue" and int(value) < 32426:
```
Comment (pradyunsg) [inline blurb/blurb.py:489]: Let's put the magic number in a variable with a clear name?

Something like: `lowest_possible_issue_number`? I don't feel strongly about the bikeshed color, as long as the name conveys something about why we're using this specific magic number.

**Review by pradyunsg — COMMENTED** [review]
A couple of non-blocking suggestions, with one of them being a nitpick that you're welcome to ignore if it's too much effort and a request to name a [magic number](https://en.m.wikipedia.org/wiki/Magic_number_(programming)).

**Comment by menkotoglou** [comment]
Thanks a lot for the review @pradyunsg. Feedback addressed already!

File: blurb/blurb.py [hunk 8]
```diff
@@ -486,6 +488,9 @@ def finish_entry():
                 # we'll complain about the *first* error
                 # we see in the blurb file, which is a
                 # better user experience.
+                if key == "gh-issue" and int(value) < lowest_possible_gh_issue_number:
+                    throw("The gh-issue number should be 32426 or above.")
```
Comment (hugovk) [inline blurb/blurb.py:492]: The force push reverted my last suggestion, so here it is again :)
```suggestion
                    throw("The gh-issue number must be 32426 or above, not a PR number.")
```

Tip: don't force push, we squash merge at the end anyway. 
(https://devguide.python.org/getting-started/pull-request-lifecycle/#quick-guide)

File: blurb/blurb.py [hunk 9]
```diff
@@ -486,6 +488,9 @@ def finish_entry():
                 # we'll complain about the *first* error
                 # we see in the blurb file, which is a
                 # better user experience.
+                if key == "gh-issue" and int(value) < lowest_possible_gh_issue_number:
+                    throw("The gh-issue number must be 32426 or above, not a PR number.")
```
Comment (pradyunsg) [inline blurb/blurb.py:492]:
```suggestion
                    throw(f"The gh-issue number must be {lowest_possible_gh_issue_number} or above, not a PR number.")
```

**Comment by menkotoglou** [comment]
Hey @pradyunsg, @hugovk, you think we're OK to merge this?

**Comment by hugovk** [comment]
Yes, let's merge. Thank you for your contribution!
